### PR TITLE
Add function name to errors and logs

### PR DESCRIPTION
**Pull request by ionut-arm**
Adding the name of the function that lead to a backend error to the log messages it generates and to the error returned to the client.

cc @ellerh

### CODE REVIEW

File: cryptoki/src/error/mod.rs [hunk 1]
```diff
@@ -131,11 +133,11 @@ impl From<std::convert::Infallible> for Error {
     }
 }
 
-impl From<RvError> for Error {
-    fn from(rv_error: RvError) -> Self {
-        Error::Pkcs11(rv_error)
-    }
-}
+// impl From<RvError> for Error {
```
Comment (wiktor-k) [inline cryptoki/src/error/mod.rs:136]: If it's not used I guess it can be removed.

File: cryptoki/src/context/general_purpose.rs [hunk 2]
```diff
@@ -120,6 +121,81 @@ pub enum Function {
     WaitForSlotEvent,
 }
 
+impl Display for Function {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
```
Comment (wiktor-k) [inline cryptoki/src/context/general_purpose.rs:126]: :thinking: quite a lot of code... is there a reason you didn't just use `write!(f, "Function::{:?}", self)`? (Maybe you didn't want to depend on the Debug trait's derive that just outputs variant's name?)

File: cryptoki/src/error/mod.rs [hunk 3]
```diff
@@ -18,7 +20,7 @@ pub enum Error {
     LibraryLoading(libloading::Error),
 
     /// All PKCS#11 functions that return non-zero translate to this error.
-    Pkcs11(RvError),
+    Pkcs11(RvError, Function),
```
Comment (wiktor-k) [inline cryptoki/src/error/mod.rs:23]: This looks like an API break...? :thinking: (nothing against it just noting for version bump).

**Review by wiktor-k — APPROVED** [review]
LGTM :+1:

File: cryptoki/src/error/mod.rs (hunk 1 above)
Comment (ionut-arm) [reply cryptoki/src/error/mod.rs:136]: Yeah, sorry, commented out when I was fixing things, forgot to delete it 😁

File: cryptoki/src/error/mod.rs (hunk 3 above)
Comment (ionut-arm) [reply cryptoki/src/error/mod.rs:23]: Hmm, indeed. I was aware, but now thinking about it, maybe I can hide the `Function` in `RvError` instead. Will need to investigate

**Comment by wiktor-k** [comment]
@ionut-arm  it seems this is largely OK to merge (after fixing conflicts). Is there anything else you'd like to add to it?

**Comment by ionut-arm** [comment]
> @ionut-arm it seems this is largely OK to merge (after fixing conflicts). Is there anything else you'd like to add to it?

Ok, I've incorporated your feedback, I don't think I can easily avoid breaking the interface since we rely on enums for errors..

**Review by wiktor-k — APPROVED** [review]
Okay, I think this looks good :+1: I'm a bit worried about extending `Error` enum but if you bump version on release it should be fine.

Thanks for this change! I think it will make debugging quite a bit easier!

**Review by gowthamsk-arm — APPROVED** [review]
Apologies for the delay. LGTM :)

**Pull request merged**